File: dotnet/src/JsonStructure/Schema/JsonStructureSchemaExporter.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

using System.Collections.Concurrent;
using System.Reflection;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;
using System.Text.Json.Serialization.Metadata;

namespace JsonStructure.Schema;

/// <summary>
/// Options for controlling JSON Structure schema generation.
/// </summary>
public sealed class JsonStructureSchemaExporterOptions
{
    /// <summary>
    /// Gets or sets the schema URI to use. Defaults to JSON Structure core v0.
    /// When <see cref="UseExtendedValidation"/> is true, this is overridden to use the extended meta-schema.
    /// </summary>
    public string SchemaUri { get; set; } = "https://json-structure.org/meta/core/v0/#";

    /// <summary>
    /// Gets or sets whether to include the $schema property. Default is true.
    /// </summary>
    public bool IncludeSchemaKeyword { get; set; } = true;

    /// <summary>
    /// Gets or sets whether to include title from XML documentation or DisplayName. Default is true.
    /// </summary>
    public bool IncludeTitles { get; set; } = true;

    /// <summary>
    /// Gets or sets whether to include description from XML documentation or Description attribute. Default is true.
    /// </summary>
    public bool IncludeDescriptions { get; set; } = true;

    /// <summary>
    /// Gets or sets whether to mark non-nullable reference types as required. Default is true.
    /// </summary>
    public bool TreatNullObliviousAsNonNullable { get; set; } = true;

    /// <summary>
    /// Gets or sets whether to use extended validation keywords from the JSON Structure Validation extension.
    /// When enabled, the schema will use the extended meta-schema URI and include a $uses clause
    /// with "JSONStructureValidation". This enables additional validation keywords such as
    /// minLength, pattern, format, minItems, maxItems, uniqueItems, contains, minProperties,
    /// maxProperties, dependentRequired, patternProperties, propertyNames, and default.
    /// Default is false.
    /// </summary>
    /// <remarks>
    /// See https://json-structure.github.io/validation/draft-vasters-json-structure-validation.html
    /// for the full specification of validation extension keywords.
    /// </remarks>
    public bool UseExtendedValidation { get; set; } = false;

    /// <summary>
    /// Gets or sets a callback to transform the generated schema.
    /// </summary>
    public Func<JsonStructureSchemaExporterContext, JsonNode, JsonNode>? TransformSchema { get; set; }
}

/// <summary>
/// Context provided to schema transformation callbacks.
/// </summary>
public sealed class JsonStructureSchemaExporterContext
{
    /// <summary>
    /// Gets the type being processed.
    /// </summary>
    public required Type Type { get; init; }

    /// <summary>
    /// Gets the property info if processing a property.
    /// </summary>
    public PropertyInfo? PropertyInfo { get; init; }

    /// <summary>
    /// Gets the JSON property info if available.
    /// </summary>
    public JsonPropertyInfo? JsonPropertyInfo { get; init; }

    /// <summary>
    /// Gets the path in the schema.
    /// </summary>
    public required string Path { get; init; }

    /// <summary>
    /// Gets whether this is the root schema.
    /// </summary>
    public bool IsRoot => string.IsNullOrEmpty(Path);
}

/// <summary>
/// Generates JSON Structure schemas from .NET types.
/// Similar to System.Text.Json.Schema.JsonSchemaExporter but for JSON Structure format.
/// </summary>
public static class JsonStructureSchemaExporter
{
    private static readonly ConcurrentDictionary<(Type, string), JsonNode> _schemaCache = new();

    /// <summary>
    /// Gets the JSON Structure schema for the specified type.
    /// </summary>
    /// <typeparam name="T">The type to generate a schema for.</typeparam>
    /// <param name="options">Optional serializer options.</param>
    /// <param name="exporterOptions">Optional exporter options.</param>
    /// <returns>A JsonNode representing the schema.</returns>
    public static JsonNode GetJsonStructureSchemaAsNode<T>(
        JsonSerializerOptions? options = null,
        JsonStructureSchemaExporterOptions? exporterOptions = null)
    {
        return GetJsonStructureSchemaAsNode(typeof(T), options, exporterOptions);
    }

    /// <summary>
    /// Gets the JSON Structure schema for the specified type.
    /// </summary>
    /// <param name="type">The type to generate a schema for.</param>
    /// <param name="options">Optional serializer options.</param>
    /// <param name="exporterOptions">Optional exporter options.</param>
    /// <returns>A JsonNode representing the schema.</returns>
    public static JsonNode GetJsonStructureSchemaAsNode(
        Type type,
        JsonSerializerOptions? options = null,
        JsonStructureSchemaExporterOptions? exporterOptions = null)
    {
        ArgumentNullException.ThrowIfNull(type);

        options ??= JsonSerializerOptions.Default;
        exporterOptions ??= new JsonStructureSchemaExporterOptions();

        // Note: Caching could be implemented here using (type, options) as key
        return GenerateSchema(type, options, exporterOptions, "", new HashSet<Type>());
    }

    private static JsonNode GenerateSchema(
        Type type,
        JsonSerializerOptions options,
        JsonStructureSchemaExporterOptions exporterOptions,
        string path,
        HashSet<Type> visitedTypes)
    {
        var schema = new JsonObject();
        var isRoot = string.IsNullOrEmpty(path);

        // Add $schema and $uses for root
        if (isRoot && exporterOptions.IncludeSchemaKeyword)
        {
            if (exporterOptions.UseExtendedValidation)
            {
                // Use extended meta-schema with validation extension
                schema["$schema"] = "https://json-structure.org/meta/extended/v0/#";
                schema["$uses"] = new JsonArray("JSONStructureValidation");
            }
            else
            {
                schema["$schema"] = exporterOptions.SchemaUri;
            }
        }

        // Handle nullable types
        var underlyingType = Nullable.GetUnderlyingType(type);
        if (underlyingType is not null)
        {
            type = underlyingType;
        }

        // Check for recursion - per JSON Structure spec, $ref must be inside type
        if (visitedTypes.Contains(type) && !type.IsPrimitive && type != typeof(string))
        {
            // Use $ref inside type property for recursive types
            schema["type"] = new JsonObject { ["$ref"] = $"#/definitions/{GetTypeName(type)}" };
            return schema;
        }

        // Map .NET type to JSON Structure type
        var structureType = GetJsonStructureType(type);
        if (structureType is not null)
        {
            schema["type"] = structureType;
        }

        // Add metadata
        if (exporterOptions.IncludeTitles)
        {
            var displayName = type.GetCustomAttribute<System.ComponentModel.DisplayNameAttribute>();
            if (displayName is not null)
            {
                schema["title"] = displayName.DisplayName;
            }
            else
            {
                schema["title"] = GetFriendlyTypeName(type);
            }
        }

        if (exporterOptions.IncludeDescriptions)
        {
            var description = type.GetCustomAttribute<System.ComponentModel.DescriptionAttribute>();
            if (description is not null)
            {
                schema["description"] = description.Description;
            }
        }

        // Handle enums
        if (type.IsEnum)
        {
            schema["type"] = "string";
            var enumValues = new JsonArray();
            foreach (var value in Enum.GetNames(type))
            {
                var jsonName = GetEnumValueName(type, value, options);
                enumValues.Add(jsonName);
            }
            schema["enum"] = enumValues;
            return ApplyTransform(schema, type, null, null, path, exporterOptions);
        }

        // Handle arrays/collections
        if (type.IsArray)
        {
            schema["type"] = "array";
            var elementType = type.GetElementType()!;
            schema["items"] = GenerateSchema(elementType, options, exporterOptions, 
                path + "/items", new HashSet<Type>(visitedTypes));
            return ApplyTransform(schema, type, null, null, path, exporterOptions);
        }

        if (type.IsGenericType)
        {
            var genericDef = type.GetGenericTypeDefinition();

            // IEnumerable<T>, List<T>, etc.
            if (IsCollectionType(genericDef))
            {
                schema["type"] = "array";
                var elementType = type.GetGenericArguments()[0];
                schema["items"] = GenerateSchema(elementType, options, exporterOptions,
                    path + "/items", new HashSet<Type>(visitedTypes));

                // Check for HashSet -> set type
                if (genericDef == typeof(HashSet<>) || genericDef == typeof(ISet<>))
                {
                    schema["type"] = "set";
                }

                return ApplyTransform(schema, type, null, null, path, exporterOptions);
            }

            // Dictionary<K,V>
            if (IsDictionaryType(genericDef))
            {
                schema["type"] = "map";
                var valueType = type.GetGenericArguments()[1];
                schema["values"] = GenerateSchema(valueType, options, exporterOptions,
                    path + "/values", new HashSet<Type>(visitedTypes));
                return ApplyTransform(schema, type, null, null, path, exporterOptions);
            }
        }

        // Handle complex objects
        if (structureType == "object" && !type.IsPrimitive && type != typeof(string))
        {
            visitedTypes.Add(type);

            var properties = new JsonObject();
            var required = new JsonArray();

            foreach (var prop in type.GetProperties(BindingFlags.Public | BindingFlags.Instance))
            {
                // Skip properties with JsonIgnore
                if (prop.GetCustomAttribute<JsonIgnoreAttribute>() is not null)
                {
                    continue;
                }

                // Get JSON property name
                var jsonName = GetJsonPropertyName(prop, options);

                // Generate property schema
                var propSchema = GenerateSchema(prop.PropertyType, options, exporterOptions,
                    path + "/properties/" + jsonName, new HashSet<Type>(visitedTypes));

                // Add property metadata
                if (propSchema is JsonObject propSchemaObj)
                {
                    if (exporterOptions.IncludeDescriptions)
                    {
                        var desc = prop.GetCustomAttribute<System.ComponentModel.DescriptionAttribute>();
                        if (desc is not null)
                        {
                            propSchemaObj["description"] = desc.Description;
                        }
                    }

                    // Validation extension keywords - only emit when UseExtendedValidation is enabled
                    if (exporterOptions.UseExtendedValidation)
                    {
                        // Determine if property is an array/collection type
                        var propType = prop.PropertyType;
                        var isArrayType = propType.IsArray || 
                            (propType.IsGenericType && IsCollectionType(propType.GetGenericTypeDefinition()));

                        // Check for range constraints (minimum/maximum/exclusiveMinimum/exclusiveMaximum)
                        var range = prop.GetCustomAttribute<System.ComponentModel.DataAnnotations.RangeAttribute>();
                        if (range is not null)
                        {
                            if (range.Minimum is not null)
                            {
                                // Check for exclusive bounds (.NET 8+)
                                if (range.MinimumIsExclusive)
                                {
                                    propSchemaObj["exclusiveMinimum"] = JsonValue.Create(Convert.ToDouble(range.Minimum));
                                }
                                else
                                {
                                    propSchemaObj["minimum"] = JsonValue.Create(Convert.ToDouble(range.Minimum));
                                }
                            }
                            if (range.Maximum is not null)
                            {
                                // Check for exclusive bounds (.NET 8+)
                                if (range.MaximumIsExclusive)
                                {
                                    propSchemaObj["exclusiveMaximum"] = JsonValue.Create(Convert.ToDouble(range.Maximum));
                                }
                                else
                                {
                                    propSchemaObj["maximum"] = JsonValue.Create(Convert.ToDouble(range.Maximum));
                                }
                            }
                        }

                        // Check for string length constraints (minLength/maxLength)
                        var stringLength = prop.GetCustomAttribute<System.ComponentModel.DataAnnotations.StringLengthAttribute>();
                        if (stringLength is not null)
                        {
                            propSchemaObj["minLength"] = stringLength.MinimumLength;
                            propSchemaObj["maxLength"] = stringLength.MaximumLength;
                        }

                        // Check for minimum length - maps to minLength for strings, minItems for arrays
                        var minLength = prop.GetCustomAttribute<System.ComponentModel.DataAnnotations.MinLengthAttribute>();
                        if (minLength is not null)
                        {
                            if (isArrayType)
                            {
                                propSchemaObj["minItems"] = minLength.Length;
                            }
                            else
                            {
                                propSchemaObj["minLength"] = minLength.Length;
                            }
                        }

                        // Check for maximum length - maps to maxLength for strings, maxItems for arrays
                        var maxLengthAttr = prop.GetCustomAttribute<System.ComponentModel.DataAnnotations.MaxLengthAttribute>();
                        if (maxLengthAttr is not null)
                        {
                            if (isArrayType)
                            {
                                propSchemaObj["maxItems"] = maxLengthAttr.Length;
                            }
                            else
                            {
                                propSchemaObj["maxLength"] = maxLengthAttr.Length;
                            }
                        }

                        // Check for regex pattern
                        var regex = prop.GetCustomAttribute<System.ComponentModel.DataAnnotations.RegularExpressionAttribute>();
                        if (regex is not null)
                        {
                            propSchemaObj["pattern"] = regex.Pattern;
                        }

                        // Check for EmailAddress attribute - maps to format: "email"
                        if (prop.GetCustomAttribute<System.ComponentModel.DataAnnotations.EmailAddressAttribute>() is not null)
                        {
                            propSchemaObj["format"] = "email";
                        }
                    }

                    propSchema = ApplyTransform(propSchemaObj, prop.PropertyType, prop, null, 
                        path + "/properties/" + jsonName, exporterOptions);
                }

                properties[jsonName] = propSchema;

                // Check if required
                if (IsPropertyRequired(prop, exporterOptions))
                {
                    required.Add(jsonName);
                }
            }

            schema["properties"] = properties;
            if (required.Count > 0)
            {
                schema["required"] = required;
            }

            visitedTypes.Remove(type);
        }

        return ApplyTransform(schema, type, null, null, path, exporterOptions);
    }

    private static string? GetJsonStructureType(Type type)
    {
        // Handle nullable
        var underlying = Nullable.GetUnderlyingType(type);
        if (underlying is not null)
        {
            type = underlying;
        }

        // Primitive types
        if (type == typeof(string)) return "string";
        if (type == typeof(bool)) return "boolean";
        
        // Integer types - use size-specific types
        if (type == typeof(sbyte)) return "int8";
        if (type == typeof(short)) return "int16";
        if (type == typeof(int)) return "int32";
        if (type == typeof(long)) return "int64";
        if (type == typeof(Int128)) return "int128";
        if (type == typeof(byte)) return "uint8";
        if (type == typeof(ushort)) return "uint16";
        if (type == typeof(uint)) return "uint32";
        if (type == typeof(ulong)) return "uint64";
        if (type == typeof(UInt128)) return "uint128";

        // Floating point - use spec names: float (32-bit), double (64-bit)
        // Note: Half (16-bit) maps to float as the closest available spec type
        if (type == typeof(Half)) return "float";
        if (type == typeof(float)) return "float";
        if (type == typeof(double)) return "double";
        if (type == typeof(decimal)) return "decimal";

        // Date/time types
        if (type == typeof(DateOnly)) return "date";
        if (type == typeof(TimeOnly)) return "time";
        if (type == typeof(DateTime)) return "datetime";
        if (type == typeof(DateTimeOffset)) return "datetime";
        if (type == typeof(TimeSpan)) return "duration";

        // Other types
        if (type == typeof(Guid)) return "uuid";
        if (type == typeof(Uri)) return "uri";
        if (type == typeof(byte[])) return "binary";
        if (type == typeof(ReadOnlyMemory<byte>)) return "binary";
        if (type == typeof(Memory<byte>)) return "binary";

        // Object for complex types
        if (type.IsClass || type.IsValueType)
        {
            return "object";
        }

        return null;
    }

    private static string GetJsonPropertyName(PropertyInfo prop, JsonSerializerOptions options)
    {
        // Check for JsonPropertyName attribute
        var attr = prop.GetCustomAttribute<JsonPropertyNameAttribute>();
        if (attr is not null)
        {
            return attr.Name;
        }

        // Use naming policy if configured
        if (options.PropertyNamingPolicy is not null)
        {
            return options.PropertyNamingPolicy.ConvertName(prop.Name);
        }

        return prop.Name;
    }

    private static string GetEnumValueName(Type enumType, string valueName, JsonSerializerOptions options)
    {
        var field = enumType.GetField(valueName);
        if (field is not null)
        {
            var attr = field.GetCustomAttribute<JsonPropertyNameAttribute>();
            if (attr is not null)
            {
                return attr.Name;
            }
        }

        if (options.PropertyNamingPolicy is not null)
        {
            return options.PropertyNamingPolicy.ConvertName(valueName);
        }

        return valueName;
    }

    private static bool IsPropertyRequired(PropertyInfo prop, JsonStructureSchemaExporterOptions options)
    {
        // Check for Required attribute
        if (prop.GetCustomAttribute<System.ComponentModel.DataAnnotations.RequiredAttribute>() is not null)
        {
            return true;
        }

        // Check for JsonRequired attribute
        if (prop.GetCustomAttribute<JsonRequiredAttribute>() is not null)
        {
            return true;
        }

        // Check nullability context if treating null-oblivious as non-nullable
        if (options.TreatNullObliviousAsNonNullable)
        {
            var context = new NullabilityInfoContext();
            var nullabilityInfo = context.Create(prop);
            if (nullabilityInfo.WriteState == NullabilityState.NotNull)
            {
                return true;
            }
        }

        return false;
    }

    private static bool IsCollectionType(Type genericDef)
    {
        return genericDef == typeof(List<>) ||
               genericDef == typeof(IList<>) ||
               genericDef == typeof(ICollection<>) ||
               genericDef == typeof(IEnumerable<>) ||
               genericDef == typeof(IReadOnlyList<>) ||
               genericDef == typeof(IReadOnlyCollection<>) ||
               genericDef == typeof(HashSet<>) ||
               genericDef == typeof(ISet<>);
    }

    private static bool IsDictionaryType(Type genericDef)
    {
        return genericDef == typeof(Dictionary<,>) ||
               genericDef == typeof(IDictionary<,>) ||
               genericDef == typeof(IReadOnlyDictionary<,>);
    }

    private static string GetTypeName(Type type)
    {
        if (type.IsGenericType)
        {
            var name = type.Name;
            var index = name.IndexOf('`');
            if (index > 0)
            {
                name = name[..index];
            }
            var args = string.Join("_", type.GetGenericArguments().Select(GetTypeName));
            return $"{name}_{args}";
        }
        return type.Name;
    }

    private static string GetFriendlyTypeName(Type type)
    {
        if (type.IsGenericType)
        {
            var name = type.Name;
            var index = name.IndexOf('`');
            if (index > 0)
            {
                name = name[..index];
            }
            var args = string.Join(", ", type.GetGenericArguments().Select(GetFriendlyTypeName));
            return $"{name}<{args}>";
        }
        return type.Name;
    }

    private static JsonNode ApplyTransform(
        JsonNode schema,
        Type type,
        PropertyInfo? prop,
        JsonPropertyInfo? jsonProp,
        string path,
        JsonStructureSchemaExporterOptions options)
    {
        if (options.TransformSchema is null)
        {
            return schema;
        }

        var context = new JsonStructureSchemaExporterContext
        {
            Type = type,
            PropertyInfo = prop,
            JsonPropertyInfo = jsonProp,
            Path = path
        };

        return options.TransformSchema(context, schema);
    }

    /// <summary>
    /// Gets the JSON Structure schema for the specified type using JsonSerializerOptions.
    /// </summary>
    /// <param name="options">The serializer options.</param>
    /// <param name="type">The type to generate a schema for.</param>
    /// <param name="exporterOptions">Optional exporter options.</param>
    /// <returns>A JsonNode representing the schema.</returns>
    public static JsonNode GetJsonStructureSchemaAsNode(
        this JsonSerializerOptions options,
        Type type,
        JsonStructureSchemaExporterOptions? exporterOptions = null)
    {
        ArgumentNullException.ThrowIfNull(options);
        ArgumentNullException.ThrowIfNull(type);

        return GetJsonStructureSchemaAsNode(type, options, exporterOptions);
    }

    /// <summary>
    /// Gets the JSON Structure schema for the specified type using JsonTypeInfo.
    /// </summary>
    /// <param name="typeInfo">The type info.</param>
    /// <param name="exporterOptions">Optional exporter options.</param>
    /// <returns>A JsonNode representing the schema.</returns>
    public static JsonNode GetJsonStructureSchemaAsNode(
        this JsonTypeInfo typeInfo,
        JsonStructureSchemaExporterOptions? exporterOptions = null)
    {
        ArgumentNullException.ThrowIfNull(typeInfo);

        return GetJsonStructureSchemaAsNode(typeInfo.Type, typeInfo.Options, exporterOptions);
    }
}

File: java/src/main/java/org/json_structure/schema/JsonStructureSchemaExporter.java
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

package org.json_structure.schema;

import com.fasterxml.jackson.annotation.*;
import com.fasterxml.jackson.databind.*;
import com.fasterxml.jackson.databind.introspect.*;
import com.fasterxml.jackson.databind.node.*;

import java.lang.reflect.*;
import java.math.BigDecimal;
import java.math.BigInteger;
import java.net.URI;
import java.time.*;
import java.util.*;

/**
 * Generates JSON Structure schemas from Java types using Jackson type introspection.
 * Similar to System.Text.Json.Schema.JsonSchemaExporter but for JSON Structure format.
 */
public final class JsonStructureSchemaExporter {

    private JsonStructureSchemaExporter() {
        // Static utility class
    }

    /**
     * Gets the JSON Structure schema for the specified type.
     *
     * @param type   the type to generate a schema for
     * @param mapper the ObjectMapper to use for introspection
     * @return a JsonNode representing the schema
     */
    public static ObjectNode getSchemaAsNode(Class<?> type, ObjectMapper mapper) {
        return getSchemaAsNode(type, mapper, null);
    }

    /**
     * Gets the JSON Structure schema for the specified type.
     *
     * @param type    the type to generate a schema for
     * @param mapper  the ObjectMapper to use for introspection
     * @param options optional exporter options
     * @return a JsonNode representing the schema
     */
    public static ObjectNode getSchemaAsNode(Class<?> type, ObjectMapper mapper,
                                              JsonStructureSchemaExporterOptions options) {
        Objects.requireNonNull(type, "type cannot be null");
        Objects.requireNonNull(mapper, "mapper cannot be null");

        if (options == null) {
            options = new JsonStructureSchemaExporterOptions();
        }

        return generateSchema(type, mapper, options, "", new HashSet<>());
    }

    private static ObjectNode generateSchema(Class<?> type, ObjectMapper mapper,
                                              JsonStructureSchemaExporterOptions options,
                                              String path, Set<Class<?>> visitedTypes) {
        ObjectNode schema = mapper.createObjectNode();
        boolean isRoot = path.isEmpty();

        // Add $schema and $uses for root
        if (isRoot && options.isIncludeSchemaKeyword()) {
            if (options.isUseExtendedValidation()) {
                // Use extended meta-schema with validation extension
                schema.put("$schema", "https://json-structure.org/meta/extended/v0/#");
                ArrayNode usesArray = schema.putArray("$uses");
                usesArray.add("JSONStructureValidation");
            } else {
                schema.put("$schema", options.getSchemaUri());
            }
        }

        // Handle wrapper types
        type = unwrapType(type);

        // Check for recursion - per JSON Structure spec, $ref must be inside type
        if (visitedTypes.contains(type) && !type.isPrimitive() && type != String.class) {
            ObjectNode refObj = mapper.createObjectNode();
            refObj.put("$ref", "#/definitions/" + getTypeName(type));
            schema.set("type", refObj);
            return applyTransform(schema, type, null, null, path, options);
        }

        // Map Java type to JSON Structure type
        String structureType = getJsonStructureType(type);
        if (structureType != null) {
            schema.put("type", structureType);
        }

        // Add metadata
        if (options.isIncludeTitles()) {
            schema.put("title", getFriendlyTypeName(type));
        }

        // Handle enums
        if (type.isEnum()) {
            schema.put("type", "string");
            ArrayNode enumValues = schema.putArray("enum");
            for (Object constant : type.getEnumConstants()) {
                String name = getEnumValueName((Enum<?>) constant, mapper);
                enumValues.add(name);
            }
            return applyTransform(schema, type, null, null, path, options);
        }

        // Handle byte[] as binary (special case before generic arrays)
        if (type == byte[].class) {
            schema.put("type", "binary");
            return applyTransform(schema, type, null, null, path, options);
        }

        // Handle arrays
        if (type.isArray()) {
            schema.put("type", "array");
            Class<?> componentType = type.getComponentType();
            schema.set("items", generateSchema(componentType, mapper, options,
                    path + "/items", new HashSet<>(visitedTypes)));
            return applyTransform(schema, type, null, null, path, options);
        }

        // Handle collections
        if (Collection.class.isAssignableFrom(type)) {
            if (Set.class.isAssignableFrom(type)) {
                schema.put("type", "set");
            } else {
                schema.put("type", "array");
            }
            // Generic type info would require more complex handling with TypeReference
            return applyTransform(schema, type, null, null, path, options);
        }

        // Handle maps
        if (Map.class.isAssignableFrom(type)) {
            schema.put("type", "map");
            // Generic type info would require more complex handling
            return applyTransform(schema, type, null, null, path, options);
        }

        // Handle complex objects
        if ("object".equals(structureType) && !type.isPrimitive() && type != String.class) {
            visitedTypes.add(type);

            ObjectNode properties = mapper.createObjectNode();
            ArrayNode required = mapper.createArrayNode();

            // Use Jackson's introspection to get properties
            JavaType javaType = mapper.getTypeFactory().constructType(type);
            BeanDescription beanDesc = mapper.getSerializationConfig().introspect(javaType);

            for (BeanPropertyDefinition prop : beanDesc.findProperties()) {
                if (prop.couldSerialize()) {
                    String jsonName = prop.getName();
                    Class<?> propType = getPropertyType(prop);

                    if (propType == null) continue;

                    // Check for JsonIgnore
                    if (isIgnored(prop)) continue;

                    // Generate property schema
                    ObjectNode propSchema = generateSchema(propType, mapper, options,
                            path + "/properties/" + jsonName, new HashSet<>(visitedTypes));

                    // Add property metadata
                    if (options.isIncludeDescriptions()) {
                        String description = getPropertyDescription(prop);
                        if (description != null) {
                            propSchema.put("description", description);
                        }
                    }

                    // Check for validation annotations
                    addValidationConstraints(prop, propSchema);

                    properties.set(jsonName, applyTransform(propSchema, propType,
                            prop.getField() != null ? prop.getField().getAnnotated() : null,
                            prop.getGetter() != null ? prop.getGetter().getAnnotated() : null,
                            path + "/properties/" + jsonName, options));

                    // Check if required
                    if (isPropertyRequired(prop, options)) {
                        required.add(jsonName);
                    }
                }
            }

            schema.set("properties", properties);
            if (required.size() > 0) {
                schema.set("required", required);
            }

            visitedTypes.remove(type);
        }

        return applyTransform(schema, type, null, null, path, options);
    }

    private static Class<?> unwrapType(Class<?> type) {
        // Unwrap Optional
        if (type == Optional.class) {
            return Object.class; // Would need generic type info
        }
        return type;
    }

    private static String getJsonStructureType(Class<?> type) {
        // Primitive types
        if (type == String.class || type == CharSequence.class) return "string";
        if (type == boolean.class || type == Boolean.class) return "boolean";

        // Integer types
        if (type == byte.class || type == Byte.class) return "int8";
        if (type == short.class || type == Short.class) return "int16";
        if (type == int.class || type == Integer.class) return "int32";
        if (type == long.class || type == Long.class) return "int64";

        // BigInteger for larger types
        if (type == BigInteger.class) return "int128"; // Or could be uint128

        // Floating point - use spec names: float (32-bit), double (64-bit)
        if (type == float.class || type == Float.class) return "float";
        if (type == double.class || type == Double.class) return "double";
        if (type == BigDecimal.class) return "decimal";

        // Date/time types
        if (type == LocalDate.class) return "date";
        if (type == LocalTime.class) return "time";
        if (type == LocalDateTime.class) return "datetime";
        if (type == OffsetDateTime.class) return "datetime";
        if (type == ZonedDateTime.class) return "datetime";
        if (type == Instant.class) return "datetime";
        if (type == Duration.class) return "duration";
        if (type == Period.class) return "duration";

        // Other types
        if (type == UUID.class) return "uuid";
        if (type == URI.class) return "uri";
        if (type == java.net.URL.class) return "uri";
        if (type == byte[].class) return "binary";

        // Object for complex types
        if (!type.isPrimitive() && !type.isEnum() && !type.isArray() &&
                !Collection.class.isAssignableFrom(type) && !Map.class.isAssignableFrom(type)) {
            return "object";
        }

        return null;
    }

    private static Class<?> getPropertyType(BeanPropertyDefinition prop) {
        AnnotatedMember member = prop.getPrimaryMember();
        if (member != null) {
            return member.getRawType();
        }
        if (prop.getField() != null) {
            return prop.getField().getRawType();
        }
        if (prop.getGetter() != null) {
            return prop.getGetter().getRawType();
        }
        return null;
    }

    private static boolean isIgnored(BeanPropertyDefinition prop) {
        AnnotatedMember member = prop.getPrimaryMember();
        if (member != null) {
            JsonIgnore ignore = member.getAnnotation(JsonIgnore.class);
            if (ignore != null && ignore.value()) return true;
        }
        return false;
    }

    private static String getPropertyDescription(BeanPropertyDefinition prop) {
        AnnotatedMember member = prop.getPrimaryMember();
        if (member != null) {
            JsonPropertyDescription desc = member.getAnnotation(JsonPropertyDescription.class);
            if (desc != null) {
                return desc.value();
            }
        }
        return null;
    }

    private static boolean isPropertyRequired(BeanPropertyDefinition prop, JsonStructureSchemaExporterOptions options) {
        // Check for JsonProperty.required
        AnnotatedMember member = prop.getPrimaryMember();
        if (member != null) {
            JsonProperty jsonProp = member.getAnnotation(JsonProperty.class);
            if (jsonProp != null && jsonProp.required()) {
                return true;
            }
        }

        // Primitive types are always required (can't be null)
        Class<?> type = getPropertyType(prop);
        if (type != null && type.isPrimitive()) {
            return true;
        }

        return false;
    }

    private static void addValidationConstraints(BeanPropertyDefinition prop, ObjectNode schema) {
        AnnotatedMember member = prop.getPrimaryMember();
        if (member == null) return;

        // Check for common validation annotations
        // Note: These would typically come from Jakarta validation annotations
        // This is a simplified implementation

        // Pattern annotation (if using regex)
        // Size annotation for collections/strings
        // Min/Max for numbers

        // For now, we'll rely on the Jackson-based metadata
    }

    private static String getEnumValueName(Enum<?> value, ObjectMapper mapper) {
        try {
            Field field = value.getDeclaringClass().getField(value.name());
            JsonProperty prop = field.getAnnotation(JsonProperty.class);
            if (prop != null && !prop.value().isEmpty()) {
                return prop.value();
            }
        } catch (NoSuchFieldException e) {
            // Fall through
        }

        // Use Jackson's naming strategy
        return value.name();
    }

    private static String getTypeName(Class<?> type) {
        if (type.isArray()) {
            return getTypeName(type.getComponentType()) + "Array";
        }
        return type.getSimpleName();
    }

    private static String getFriendlyTypeName(Class<?> type) {
        if (type.isArray()) {
            return getFriendlyTypeName(type.getComponentType()) + "[]";
        }
        String name = type.getSimpleName();

        // Handle generic types
        TypeVariable<?>[] typeParams = type.getTypeParameters();
        if (typeParams.length > 0) {
            StringBuilder sb = new StringBuilder(name);
            sb.append("<");
            for (int i = 0; i < typeParams.length; i++) {
                if (i > 0) sb.append(", ");
                sb.append(typeParams[i].getName());
            }
            sb.append(">");
            return sb.toString();
        }

        return name;
    }

    private static ObjectNode applyTransform(ObjectNode schema, Class<?> type,
                                              Field field, Method method, String path,
                                              JsonStructureSchemaExporterOptions options) {
        if (options.getSchemaTransformer() == null) {
            return schema;
        }

        JsonStructureSchemaExporterContext context =
                new JsonStructureSchemaExporterContext(type, field, method, path);

        return options.getSchemaTransformer().transform(context, schema);
    }
}
